lib/logstash/inputs/syslog.rb

```diff
@@ -107,13 +108,32 @@ def run(output_queue)
   end # def run
 
   private
-  def udp_listener(output_queue)
-    @logger.info("Starting syslog udp listener", :address => "#{@host}:#{@port}")
+  def udp_server(output_queue)
+    udp_listener(output_queue)
+  rescue => e
+    if @shutdown_requested.false?
+      @logger.warn("syslog udp listener died", :address => "#{@host}:#{@port}", :exception => e, :backtrace => e.backtrace)
+      sleep(5)
+      retry
+    end
+  end
 
-    if @udp
-      @udp.close
+  private
+  def tcp_server(output_queue)
+    tcp_listener(output_queue)
+  rescue => e
+    if @shutdown_requested.false?
+      @logger.warn("syslog tcp listener died", :address => "#{@host}:#{@port}", :exception => e, :backtrace => e.backtrace)
+      sleep(5)
+      retry
     end
+  end
```

Comment:
agree, I will refactor both decode blocks and wrap decoding exception so that they are excluded in the upper server rescue + retry and the retry logic only happen on socket exceptions.